core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -314,9 +315,16 @@ class MetadataCache(brokerId: Int) extends Logging {
           error(s"Listeners are not identical across brokers: $aliveNodes")
       }
 
+      val newTopicIds = updateMetadataRequest.topicStates().asScala
+        .map(topicState => (topicState.topicName(), topicState.topicId()))
+        .filter(_._2 != Uuid.ZERO_UUID).toMap
+      val topicIds = mutable.Map.empty[String, Uuid]
+      topicIds.addAll(metadataSnapshot.topicIds)
+      topicIds.addAll(newTopicIds)
```

Comment:
We seem to be adding to the topicIds/topicNames map, but never removing anything.

Reply:
Hi, @rajinisivaram ,Thank you for your reply. I find that when KafkaController deletes a topic it will send `updateMetadataRequest` with all the partitions in it, and set `state.leader=-2` indicating the partition is deleted. And the `MetadataCache` will not remove the topic from cache but just keep the `UpdateMetadataPartitionState` of the deleted topics, so I also keep the topicId when deleted.

here I tried to delete a topic and then describe it.

➜  ~ kafka-topics.sh --delete --zookeeper localhost:2181 --topic old-version-topic
Topic old-version-topic is marked for deletion.
Note: This will have no impact if delete.topic.enable is not set to true.

➜  ~ kafka-topics.sh --describe --zookeeper localhost:2181 --topic old-version-topic
Topic: old-version-topic TopicId: wRPl6VAlQeyE77bDxEESzg  PartitionCount: 1	ReplicationFactor: 1 Configs: 	MarkedForDeletion: true
	Topic: old-version-topic	Partition: 0	Leader: 0	Replicas: 0	Isr: 0	MarkedForDeletion: true

So the topicId of the deleted topic should also remain when deleted, when do you think we should delete the topicId from MetadataCache? suggestions are welcomed.

Reply:
When a topic is deleted, brokers process UpdateMetadataRequest and remove deleted topics from their cache. We track deletion state in ZooKeeper and as you mentioned, you can get this information by directly going to ZK in kafka-topics.sh. But we don't retain that information in every broker. I would remove topic id in the code segment just below this when the topic is removed from the MetadataCache since we cannot clearly have a map that keeps growing in brokers. Is there a reason why we would want to retain topic id in every broker even after the topic has been deleted? We can't get this information through existing metadata request from brokers anyway. I guess in future, we can add additional metadata to track deleted topic ids if we wanted to, but for now it seems better to delete topic ids from MetadataCache when we delete the topic from the cache.  What do you think?

Reply:
So it seems like there is some logic to remove the partition states of deleted topics from the MetadataSnapshot. Would we want to do something similar there? Apologies if I'm missing something

Reply:
Yes, that was my suggestion too, but my wording wasn't right. I will reword that to avoid confusion.

Reply:
Thank you for your suggestions, I now understand the solution.